### ISSUE TITLE
Incorrectly reporting failing tests and not running all tests..

### ISSUE THREAD

**Issue by dotnetprofessional** [wallabyjs/public#1028]
### Issue description or question
I've configured Walaby.js to run on my OSS project and I'm getting incorrect results. You can find the project I'm running against here:

https://github.com/dotnetprofessional/ABusJS

The code is the same as I'm using, minus the Wallaby config.

When running the tests npm test, from the command line I get the following results:

```javascript
Test Suites: 2 skipped, 8 passed, 8 of 10 total
Tests:       7 skipped, 105 passed, 112 total
```
However Wallaby.js reports the following:
​​​​
```javascript
1 failing tests, 59 passing​​​​​
```

Looking at the code that is failing:

```javascript
TypeError: Class constructor TestMessageHandler cannot be invoked without 'new'​​
  at Suite.describe ​Tests/ABus.Subscription.Spec.ts:201:8​
  at Suite.Object.<anonymous>.describe ​Tests/ABus.Subscription.Spec.ts:198:4​
  at Object.<anonymous> ​Tests/ABus.Subscription.Spec.ts:196:0​
```

​​
It seems Wallaby.js maybe having issues with a class that has a decorator on it. These tests are validating that the custom decorators work.

The other thing is that even though those tests are failing, it seems to be preventing the running of many other tests, as its missing nearly half of the available tests. I'd expect a failing test not to prevent others from running.
  
### Wallaby.js configuration file

```javascript
module.exports = function (w) {
    return {
        files: [
            'app/**/*.ts'
        ],
        tests: [
            'Tests/**/*.ts'
        ],
        env: {
            type: 'node',
            runner: 'node'
        },
        // or any other supported testing framework:
        // https://wallabyjs.com/docs/integration/overview.html#supported-testing-frameworks
        testFramework: 'jest',
        debug: true
    };
};
```

### Code editor or IDE name and version
Visual Studio Code v1.9.1

### OS name and version
Windows 10

**Comment by ArtemGovorov** [comment]
Actually wallaby had correctly found and reported a bug for you.

When you compile your code to ES2016, as your `tsconfig.json` targets it, the code like this is produced:

```javascript
let TestMessageHandler = class TestMessageHandler {
};
TestMessageHandler.apply();
```

I have simplified it, but the idea is similar because your [decorator does eventually call apply on the passed class function](https://github.com/dotnetprofessional/ABusJS/blob/1de6349601cb5a7be3c6099e559e5a7e2206424c/app/Decorators.ts#L65).

If you run this simple code, you'll get the `Class constructor TestMessageHandler cannot be invoked without 'new'​​` and it is expected, because ES2016 doesn't support class constructor invocations without `new`.

Now if the same TypeScript code is transpiled to ES5, you'll get the valid and running code:

```javascript
var TestMessageHandler = (function () {
  function TestMessageHandler() {

  }
  return TestMessageHandler;
}());

TestMessageHandler.apply();
```

Now why does it work when you run `npm test`? The answers is - your `tsconfig.json` file is simply ignored. You are using `ts-jest` module to help Jest transpile TypeScript and are [specifying the setting for the module in your `package.json`](https://github.com/dotnetprofessional/ABusJS/blob/1de6349601cb5a7be3c6099e559e5a7e2206424c/package.json#L44).

The problem is that when `__TS_CONFIG__` for `ts-jest` is specified, it [doesn't even load the `tsconfig.json`](https://github.com/kulshekhar/ts-jest/blob/master/src/utils.ts#L72). Not sure if it is a big or not, but your TypeScript code is transpiled completely differently from what you'll have in production build.

If you remove the setting
```
    "globals": {
      "__TS_CONFIG__": {
        "module": "commonjs",
        "jsx": "react"
      }
    }
```
from your `package.json`, and remove all comments from your `tsconfig.json` (because `ts-jest` doesn't support comments in the `tsconfig` file), and run 
```
npm test -- --no-cache
```

you will make `ts-jest` to use your `tsconfig.json`, and you'll get the same errors.

Also, you may notice that the  reported number of tests is slightly different from your real test number. 

Here is why it happens (in both Jest and Wallaby). Also, there were few smaller issues in your wallaby config that I fixed: the config didn't have [`files` correctly specified](https://wallabyjs.com/docs/config/files.html), didn't [use your Jest config from `package.json`](https://wallabyjs.com/docs/integration/jest.html#jest-configuration), etc. These issues were also affecting the number of tests being run/reported.

Your test file has this code:

```javascript
    describe("using a string literal to define the messageFilter", () => {
        var bus = new Bus().makeGlobal();
        // Provide an instance for the handler to attach to
        let handlerClass = new TestMessageHandler(10);

        it("should register subscriber for the message type", async () => {
            expect(bus.subscriberCount(testData.TestMessage.TYPE)).toBe(1);
        });

        it("should call handler with correct class instance", async () => {
            bus.sendAsync(new testData.TestMessage(""));
            expect(handlerClass.value).toBe(110);
        });
    });
```

Let's simplify it to:
```javascript
    describe("...", () => {

        throw new Error('...')

        it("...", () => {
            ...
        });

        it("...", () => {
            ...
        });

        it("...", () => {
            ...
        });
       ...
    });
```
or even easier, try running this in Jest:
```javascript
    throw new Error('...')
    describe("...", () => {
       ...
    });
```


When Jest is loading your tests, it hits some code that can potentially raise an error and when it does, Jest simply can't even read the rest of the tests in that block. So It doesn't report them, it doesn't even know they are there. To avoid issues like this, I'd recommend not placing test initialisation code outside `it` or hooks [like `beforeEach`](https://facebook.github.io/jest/docs/api.html#beforeeachfn). Because when the exception is raised from a test or a `beforeEach` hook, Jest is able to catch it and keep executing other tests. 

So the issue with the reported test number is fundamentally Wallaby specific, but is specific to the way Jest and many other testing frameworks work. Having said that Wallaby in some cases does stop running tests when it hits declaration issues like this - that may badly affect other tests. We are going to add a setting to allow wallaby to keep running tests in such cases, but I'm not sure if the setting is going to suit everyone.

Now back to the `Class constructor TestMessageHandler cannot be invoked without 'new'​​` issue and how to fix it. It depends on how do you plan to build your application:
- If you'd like to target ES2016, then you [have a bug in your decorator](https://github.com/dotnetprofessional/ABusJS/blob/master/app/Decorators.ts#L65) that you'll need to fix.
- If you are ok to target ES5, then you'll need to change your `tsconfig.json` - `"target": "es5"`.

If you choose option 1, you may also want to consider configuring `ts-jest` better to avoid obscuring such issues.

Here's the working (if you change your `tsconfig.json` to `"target": "es5"`) wallaby config for you:
```javascript
module.exports = function (w) {
  return {
    files: [
      'package.json',
      'app/**/*.ts',
      'Tests/**/*.+(ts|json|txt)',
      '!Tests/**/*.Spec.ts'
    ],
    tests: [
      'Tests/**/*.Spec.ts'
    ],
    env: {
      type: 'node',
      runner: 'node'
    },
    testFramework: 'jest',
    setup: function (wallaby) {
      wallaby.testFramework.configure(require('./package.json').jest);
    },
    debug: true
  };
};
```

**Comment by ArtemGovorov** [comment]
A couple more notes that I forgot to mention in the above post:
- wallaby is not using `ts-jest`, it compiles TypeScript itself (using your `tsconfig.json`),
- ES5 is the default `target`, so when `ts-jest` doesn't load `tsconfig.json`, ES5 target is used.
- in the case you'd like to keep your config targeting ES2016 in the `tsconfig`, but don't want to address the issue in code for some reason, and just need wallaby to make the tests pass (and work like `ts-jest`), you may adjust [TS compiler options](https://wallabyjs.com/docs/integration/typescript.html#typescript-compiler-settings) right in the wallaby config:
```javascript
module.exports = function (wallaby) {
  const compilerOptions = require('./tsconfig.json').compilerOptions;
  compilerOptions.target = 'es5';
  return {
    ...
    compilers: {
      '**/*.ts?(x)': wallaby.compilers.typeScript(compilerOptions)
    },
    ...
  };
};
```

**Comment by dotnetprofessional** [comment]
Thanks for the very detailed explanation! I've change the code to target ES5 as I needed to do that anyway.  I really appreciate the additional info you provided, I'm still fairly new to JavaScript in general.

Your changes have indeed allowed me to see the bug using npm-test, though I've opted for the ES5 to avoid this issue (still good to know I had a bug). I did have one little issue, seems one of my tests used a lower case spec so it was being ignored, but I found it. So I have them all running now.

Thanks a lot for your help!
